src/Raven.Server/Documents/TcpHandlers/SubscriptionConnection.cs

```diff
@@ -570,61 +612,95 @@ private async Task ProcessSubscriptionAsync()
 
-                                using (docsContext.OpenReadTransaction())
-                                {
-                                    var globalEtag = useRevisions ?
-                                        TcpConnection.DocumentDatabase.DocumentsStorage.RevisionsStorage.GetLastRevisionEtag(docsContext, Subscription.Collection) :
-                                        TcpConnection.DocumentDatabase.DocumentsStorage.GetLastDocumentEtag(docsContext.Transaction.InnerTransaction, Subscription.Collection);
+                                        if (globalEtag > _startEtag)
+                                            continue;
+                                    }
 
-                                    if (globalEtag > _startEtag)
+                                    AssertCloseWhenNoDocsLeft();
+                                    
+                                    if (await WaitForChangedDocuments(replyFromClientTask)) 
                                         continue;
                                 }
+                            }
 
-                                AssertCloseWhenNoDocsLeft();
-
-                                if (await WaitForChangedDocuments(replyFromClientTask))
-                                    continue;
+                            // using (batchParentScope.For(SubscriptionBatchOperations.AckScope))
+                            using (batchParentScope.For(SubscriptionOperationScope.Ack))
+                            {
+                                batchParentScope.RecordStartWaitingForClientAck(DateTime.UtcNow);
+                                
+                                (replyFromClientTask, subscriptionChangeVectorBeforeCurrentBatch) =
+                                    await WaitForClientAck(replyFromClientTask, subscriptionChangeVectorBeforeCurrentBatch);
+
+                                batchParentScope.RecordClientAckTime(Stats.LastAckReceivedAt);
+                                    
+                                updateBatchPerformanceStats();
                             }
                         }
+                        catch (Exception e)
+                        {
+                            batchParentScope.RecordException(e.Message);
+                            throw;
+                        }
                     }
-
-                    (replyFromClientTask, subscriptionChangeVectorBeforeCurrentBatch) =
-                        await WaitForClientAck(replyFromClientTask, subscriptionChangeVectorBeforeCurrentBatch);
                 }
 
                 CancellationTokenSource.Token.ThrowIfCancellationRequested();
             }
         }
 
+        private void updateBatchPerformanceStats()
+        {
+            _lastBatchStats.Complete();
+            AddBatchPerformanceStatsToBatchesHistory(_lastBatchStats);
+            TcpConnection.DocumentDatabase.RaiseEndBatchNotification(_options.SubscriptionName, _lastBatchStats);
```

Comment:
TcpConnection.DocumentDatabase.RaiseEndBatchNotification

If you would see such code... Can you determinate what kind of batch was finished? I'm not, so either move those events to thin glike: DocumentDatabase.Subscriptions.RaiseEndBatchNotification or add word subscription to its name.

Reply:
renaming to an easier to read form:

```
RaiseNotificationForSubscriptionTaskAdded
RaiseNotificationForSubscriptionTaskRemoved
RaiseNotificationForSubscriptionConnectionEnded
RaiseNotificationForSubscriptionBatchEnded
```

Reply:
like @arekpalinski said move this to subscription storage and drop word subscription from name to avoid repeating yourself

Reply:
done